### PR TITLE
Optimise rejection of out-of-range `long` values

### LINKED ISSUE #40323: Feeding non-numeric data into a `long` field may consume excessive CPU
I've seen a couple of cases where all the `write` threads are heroically trying to parse some very large numbers on the way into a `long` field, consuming far too much CPU in the process. The hot threads output reported many threads stuck doing `BigInteger` shenanigans such as this:

```
       java.math.BigInteger.square(BigInteger.java:1899)
       java.math.BigInteger.squareToomCook3(BigInteger.java:2053)
       java.math.BigInteger.square(BigInteger.java:1899)
       java.math.BigInteger.squareToomCook3(BigInteger.java:2051)
       java.math.BigInteger.square(BigInteger.java:1899)
       java.math.BigInteger.squareToomCook3(BigInteger.java:2051)
       java.math.BigInteger.square(BigInteger.java:1899)
       java.math.BigInteger.squareToomCook3(BigInteger.java:2053)
       java.math.BigInteger.square(BigInteger.java:1899)
       java.math.BigInteger.pow(BigInteger.java:2306)
       java.math.BigDecimal.bigTenToThe(BigDecimal.java:3543)
       java.math.BigDecimal.bigMultiplyPowerTen(BigDecimal.java:3676)
       java.math.BigDecimal.setScale(BigDecimal.java:2445)
       java.math.BigDecimal.toBigInteger(BigDecimal.java:3025)
       org.elasticsearch.common.xcontent.support.AbstractXContentParser.toLong(AbstractXContentParser.java:195)
       org.elasticsearch.common.xcontent.support.AbstractXContentParser.longValue(AbstractXContentParser.java:220)
       org.elasticsearch.index.mapper.NumberFieldMapper$NumberType$7.parse(NumberFieldMapper.java:679)
       org.elasticsearch.index.mapper.NumberFieldMapper$NumberType$7.parse(NumberFieldMapper.java:655)
       org.elasticsearch.index.mapper.NumberFieldMapper.parseCreateField(NumberFieldMapper.java:996)
       org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:297)
```

One example was at https://discuss.elastic.co/t/high-cpu-usage-in-elasticsearch-nodes/161504 and another was with a customer.

We fall back to `BigInteger` if a naive conversion to `long` fails, for instance if using scientific notation, but then reject the result if it doesn't fit into a `long`:

https://github.com/elastic/elasticsearch/blob/92b2e1a2091ce30d6267029b7b7b19d4b6690d64/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java#L157-L180

This means that if you try and insert a short string such as `"1e99999999"` into a long field then we spend a lot of resources converting this into a very large `BigInteger` before deciding it's too big. I think we should not try so hard to parse values like `"1e99999999"` as a `long`.

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today if you try and insert a very large number like `1e9999999` into a long
field we first construct this number as a `BigDecimal`, convert this to a
`BigInteger` and then reject it because it is out of range. Unfortunately
making such a large `BigInteger` is rather expensive.

We can avoid this expense by performing a (weaker) range check on the
`BigDecimal` representation of incoming `long`s too.

Relates #26137
Closes #40323

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/NumbersTests.java [hunk 1]
```diff
@@ -27,19 +28,26 @@
 
 public class NumbersTests extends ESTestCase {
 
+    @Timeout(millis = 10000)
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/common/NumbersTests.java]: does it take that long?

File: server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java [hunk 2]
```diff
@@ -367,17 +368,20 @@ public void testEmptyName() throws IOException {
         }
     }
 
+    @Timeout(millis = 30000)
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java]: Is it taking that long?

**Review by jpountz — APPROVED** [review]
Good catch. The change looks good to me. I'm wondering whether the timeouts on tests are left-overs?

**Comment by DaveCTurner** [comment]
I was unsure about whether to include these timeouts. The tests complete in milliseconds today with the fix, but I'm nervous about short timeouts adding to CI flakiness. Without the fix or the timeout these tests take a very long time to fail which I think isn't great either. I don't feel strongly, however, and will take them out if you think it better.
